Button onHoverIn and onHoverOut events

swese44 opened the pull request.

msftclas commented:
@swese44,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

unindented commented on src/components/Button/button.test.tsx:
```diff
@@ -163,22 +163,42 @@ describe('<Button />', () => {
     beforeEach(() => {
       clicked = false;  
+      hoveredIn = false;  
+      hoveredOut = false;
```
Why not use spies (`jest.fn()`) instead of functions that change a boolean?

swese44 replied:
fixed thanks

benlakey commented on src/components/Button/button.test.tsx:
```diff
@@ -163,22 +163,42 @@ describe('<Button />', () => {
+      hoveredOut = false;
       component = shallow(
-        <Button text={sampleText} onClick={clickMe} />,
+        <Button text={sampleText} onClick={onClick} onHoverIn={onHoverIn} onHoverOut={onHoverOut} />,
```
Consider batching up the props in an object and then using spread to make it more concise as props are added:

```javascript
<Button {...props} />
```

unindented replied:
But then you don't really know which props your forwarding. I'm more comfortable being explicit here.

benlakey replied:
Its typescript?

unindented replied:
Ha, you're right.

benlakey commented on src/components/Button/button.test.tsx:
```diff
@@ -163,22 +163,42 @@ describe('<Button />', () => {
 
     it('triggers its onClick callback', () => {
       component.simulate('click');
       expect(clicked).toBe(true);
```
In combination with the feedback from @unindented  (spies), recommend leveraging `describe` to describe scenarios and `it` to specify expectations:

```javascript
describe('when the component is clicked', function() {
  beforeEach(function() {
    component.simulate('click');
  });

  it('triggers the onClick callback', function() {
    expect(onClick).toHaveBeenCalled();
  });
});

benlakey replied:
same comments for other tests below

swese44 replied:
added thanks

annr commented:
We should probably update the README for Button as well.

annr commented on src/components/Button/button.test.tsx:
```diff
@@ -163,21 +163,50 @@ describe('<Button />', () => {
+      });
+
+      it('triggers its onHoverOut callback', () => {
+        expect(onHoverIn).not.toHaveBeenCalled();
```
Testing onHoverIn/onHoverOut "not called" is not necessary. We're using onMouseLeave/onMouseEnter on the component, and you're testing what those native attributes do. Since one handler is passed, it isn't a valuable test for us.

swese44 replied:
cool removed

swese44 commented:
* Changed onHoverIn and onHoverOut to onMouseEnter and onMouseLeave, this should be more intuitive with the standard event names
* Added onFocus and onBlur so TriState buttons can respond to keyboard events
* Added font-family to button (it was defaulting to the browser's default font-family - we can remove $fontStack once we wrap our app in a Fabric component
* Updated React peer-dependency to be flexible so we don't block YamJS (or possibly SharepointWebParts) from updating its version of React

swese44 commented:
@annr could you review again? also you should be able to set your github creds just for a single repo
